### ISSUE TITLE
Compiler Warnings

### ISSUE THREAD

**Issue by WaynePhillipsEA** [twinbasic/twinbasic#208]
Here I want to discuss compilation warnings.   At this moment in time (v0.10.1389), twinBASIC does not produce **_any warnings whatsoever_**.  Not a single one.   It does of course identify and produce errors in the VS Code UI, but the compiler can, and should, be smarter than that, spotting common mistakes and other questionable behaviour that should be reported to the user via a warning.

I propose that each type of warning should have an ID attributed to it that will never change, and we need a means of being able to control each warning individually at the project level, and perhaps the component level (via attributes).  I anticipate that each individual type of warning will be in one of three states at compilation time:
  
   1. ignored    (not reported in the UI)
   2. warning    (default state, reported as a warning in the UI, but compilation still succeeds)
   3. enforced    (the warning is upgraded to an error)

[the wording / terminology can be tweaked]

I anticipate that Option Strict will then simply toggle a group or subset of these warnings to their 'enforced' state.

Let's start by discussing what types of things we want twinBASIC to identify as **_warnings_**.

**Comment by bclothier** [comment]
I would point out that warnings should be considered as a part of the analysis extension proposal (#51). That way, we have an uniform interface for defining hints, warnings, and errors both by tB itself and by any custom analyzer we may write. 

For that reason, it may be useful to study how .NET Roslyn analyzers are implemented and use that as a starting point. The other advantage is that the analyzers then are standalone and thus can be managed via future package manager (c.f. #16). 

That said, I think those are the most common mistakes that would be helped greatly by such static analysis:

0) failing to assign to a variable, particularly an object and then using it. 
1) suspicious casting
2) late-bound calls (this is very very easy to miss and I can see a legacy code file lighting up like a Christmas tree but it needs to be pointed out)
3) implicit access to default member or to `appobject`'s member or to `Evaluate` methods
4) usage of bang operators
5) shadowing declarations

**Comment by mansellan** [comment]
I'd recommend following Microsoft's lead, by prefixing the numeric ID with a couple of alpha chars (presumably `TB` :-) ), if twinBASIC takes over the world, it'd make the error/warning _much_ easier to search for.

Also C# [categorises warnings into several levels](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/) - the advantage here is that it's much easier to suppress a whole swathe of "trivial" warnings with a single project-level switch. Not sure we'd need 5 levels though!

**Comment by mansellan** [comment]
Also, we could allow distinct sections of code to have specific suppressions using pragmas or similar.

```
#Disable TB0001
' Some Code
#Restore TB0001
```

**Comment by Kr00l** [comment]
1. The "+" operator should not be used to concenate strings. (E.g. the "-" operator raises a type mismatch). So a warning/hint should be raised to encourage using the "&" operator. Though the warning/hint should not block compilation.

2. Function/Property/Dim w/o a data type. It's clear that it defaults to Variant. But a hint should encourage to explicitly declare "As Variant" to make the intent clear.

**Comment by mansellan** [comment]
1. Use of `DefType`
2. Use of `GoSub`
3. Use of line numbers (not labels)

Listing all these separately as they should be individually suppressible as required (could be a specific use-case, or a legacy codebase with more of one construct than the others).

**Comment by WaynePhillipsEA** [comment]
Thanks everyone - all good suggestions.

> I would point out that warnings should be considered as a part of the analysis extension proposal (#51). That way, we have an uniform interface for defining hints, warnings, and errors both by tB itself and by any custom analyzer we may write.
> 
> For that reason, it may be useful to study how .NET Roslyn analyzers are implemented and use that as a starting point. The other advantage is that the analyzers then are standalone and thus can be managed via future package manager (c.f. #16).

I totally agree with this, however we have to consider that the analysis extension API is much lower down the priority list, and most of these warnings would be very simple to implement inside the compiler right now.  So I'm considering adding these warnings now, with the expectation that I will refactor them into analysis extensions when the API is available.

> Also, we could allow distinct sections of code to have specific suppressions using pragmas or similar.
> 
> ```
> #Disable TB0001
> ' Some Code
> #Restore TB0001
> ```

I was considering using attributes for changing each warning state, as they would then be attached/scoped to the component without the need to preserve/restore the state afterwards.  Something more along the lines of this:

```
[IgnoreWarnings(TB0001,TB0052)]
Public Class MyClass1
End Class

' TB0001 and TB0052 are back to their previous state here
[EnforceErrors(TB0064)]
Public Class MyClass2
End Class

' If TB0044 is ignored at the project level, it can still be switched back on for an individual component
[EnforceWarnings(TB0044)]
Public Class MyClass3
End Class
```

Thoughts?

**Comment by bclothier** [comment]
I would much prefer attribute over `#prgama` directives for the reasons [I cited earlier in the attribute issue](https://github.com/WaynePhillipsEA/twinbasic/issues/25#issuecomment-821741827). However, the attributes should always follow the same scoping rules since they are technically still code. Therefore, the proposal would violate that and we end up with non-comment `#pragma` which we don't want.

I think we want this instead:

```
'The MyClass1 will ignore warning TB0001 and TB0052
[IgnoreWarnings(TB0001,TB0052)]
Public Class MyClass1

   ' TB0001 and TB0052 are back to their previous state here
   [EnforceErrors(TB0064)]
   Public Sub Foo
   End Class
End Class

'The warning TB0001 and TB0052 are in effect again
Public Class MyClass2
End Class

' If TB0044 is ignored at the project level, it can still be switched back on for an individual component
[EnforceWarnings(TB0044)]
Public Class MyClass3
End Class
```

Respecting the scope rules would help making it consistent and easy to understand, IMO.

**Comment by WaynePhillipsEA** [comment]
@bclothier apologies if I wasn't clear.  tB attributes are always scoped to a component / member etc.  

The only difference I see between yours and my proposal is that your class scoped IgnoreWarnings attribute is not applied to the class member Foo.  That doesn't sit quite right with me as my mind would naturally expect the IgnoreWarnings attribute on the class to also apply to inner class members, sort of inherited.

**Comment by bclothier** [comment]
@WaynePhillipsEA  I misunderstood the intention. I thought that the `TB0064` covered both `TB0001` and `TB0052`. I now see that you mean them to be all independent so yes, my sample only makes sense if the `TB0064` was somehow disabled at the project level which would then cascade down to the `MyClass1`'s `Foo` member. I think we're in agreement here. Thanks for clarifying!

**Comment by WaynePhillipsEA** [comment]
As of v0.10.3604, the internal framework for supporting compiler warnings is in place and we have our first two compiler warnings, TB0001 and TB0002.   The TBxxxx warning codes will never change once assigned.

In the project Settings you can adjust each individual warning into an ERROR or IGNORE state to control how the warnings are treated during compilation of your project.

<img width="641" alt="tbCompilerWarnings1" src="https://user-images.githubusercontent.com/25738109/127448058-5fb15cfb-29f7-40a7-8508-d1f3324e169e.png">

And we have the following attributes that you can use on Class / Modules / procedures to refine how each warning is treated at a more granular level:

```
[ IgnoreWarnings(TB0001) ]
[ EnforceErrors(TB0001) ]
[ EnforceWarnings(TB0001) ]
```
You can specify multiple warning codes by separating with commas, e.g. EnforceErrors(TB0001, TB0002).

Inheritance rules apply as follows:   Procedure attributes > Class/Module attributes > Project settings.  So for example a procedure attribute of EnforceWarnings(TB0001) would override an IgnoreWarnings(TB0001) attribute on the parent Class.

**Comment by WaynePhillipsEA** [comment]
In v0.10.3664, added warning TB0003 'datatype has not been explicitly declared so will default to Variant'

**Comment by Kr00l** [comment]
I like the bang operator, especially for Recordsets. Hopefully that will be not a warning by default soon.. :E

**Comment by dmrvb** [comment]
Implied or automation type conversation could be included as its potentially poor programming.
This would have to be an option as sometimes its fine to make use of it.

**Comment by Kr00l** [comment]
Just a quick question for completeness. Would it be possible to visually represent a new category, such as "Hint". It would be between "Ignore" and "Warning". Not highlighting/disturbing in code but listed softly in the Problem list. Just asking if VS code has such an possibility.
Because maybe in future some compiler checks may be more appropriate as Hint. (Maybe)

**Comment by WaynePhillipsEA** [comment]
@Kr00l, yes VS Code supports that.  The LSP supports severity levels of Error/Warning/Information/Hint:
https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#diagnostic

I'll get the extra severity levels added to tB, as they will be useful in future.

**Comment by dmrvb** [comment]
Can I request that a Redim without a Dim generates an 0003 warning?
A simple Typo will result in accidentally getting an array of variants:......

Dim AAA() as long
...
...
Redim AA(20)       'this should get a warning

**Comment by WaynePhillipsEA** [comment]
@dmrvb Yes, looks like that one got missed.  I'll get that added for you, thanks.

**Comment by WaynePhillipsEA** [comment]
@dmrvb that's done for you in v0.10.5474.

**Comment by WaynePhillipsEA** [comment]
As per #443, tB should output a warning when a variable is determined not to be used

**Comment by Kr00l** [comment]
Perhaps 'unused variable/const' is a good candidate to be as "information" instead of "warning" by default.
Because sometimes a module has a "stock" of consts which are intentionally not "all" used. Having them as "warning" would annoy. "Information" may be softer.

**Comment by wqweto** [comment]
Unused variables are usually dealt with not by the compiler but a separate linter and same for reformatting code etc. not compilation tasks.

**Comment by WaynePhillipsEA** [comment]
Indeed, and we could argue the same for all warnings, but at this stage we don't have a separate process for this.  Provided the analysis is relatively cost-free (in terms of performance), and in this case it is, then I'm willing to include it now and migrate it later once we have an official analysis API in place (#51).

**Comment by wqweto** [comment]
MZ-Tools linter checks for unused member vars, enums, types, API declares and routines only in private scope while missing is public symbols analysis i.e. public symbols in a BAS module are available in project scope only so it's possible to find unused one. Public symbols in a CLS file on a non-ActiveX project should be inspectable too.

**Comment by mburns08109** [comment]
Perhaps we should have 4 levels of compiler messages:
Error - fails to compile
Warning - compiles, but may not produce anticipated results
Information - this is the level I would put unused variables and "magic-value-in-code" notices
Hint - future analysis API stuff may go here

**Comment by WaynePhillipsEA** [comment]
v0.13.56 added compiler warning TB0008 'Suspicious call' (#598)
